homeassistant/components/mqtt/client.py

```diff
@@ -695,6 +697,12 @@ def _mqtt_handle_message(self, msg: mqtt.MQTTMessage) -> None:
         subscriptions = self._matching_subscriptions(msg.topic)
 
         for subscription in subscriptions:
+            if (init_status := self._retained_init.get(subscription)) is None:
+                self._retained_init[subscription] = msg.retain
+            if msg.retain and init_status is True:
```

Comment:
This is not right.

If we are already subscribed to topic Foo
And we get another subscription request for Foo (which we will - say for entities sharing the availability topic)

If between when we create the Subscription object and when we actually subscribe (especially with debouncing - https://github.com/home-assistant/core/pull/88862) we receive a message on that topic (which will have RETAIN=0) we will set self._retained_init[subscription] = False

However, this counts as being initialized (sets init_status = True) as your init_status is checking to see only if it gets back a None from the get() call - not False.  

This means, when the actual subscription is sent again, and then MQTT sends back a RETAIN=1 message, the new subscription will NOT get that retain message, because it already got a RETAIN=0 message, which initialized it's entry in self._retained_init[subscription] (to False, but False != None) and so this condition will fail.

Should probably have `init_status := self._retained_init.get(subscription) or False`.   Sure, it means that if you never get a RETAIN=1 message on a topic we will keep treating it as if it was never initialized, but that doesn't really matter, because you will skip the next check because msg.retain = False anyway.

More importantly, this will fail more drastically on wildcard topics.  When you subscribe to a wildcard topic, you will get multiple (indeed, MANY) messages that ALL have RETAIN=1.  But they will each be for a different msg.topic - but they are all for the same subscription (the subscription.topic is wildcarded).

This means, with this change, you will drop all but the first initial message sent for a wildcard subscribe.

Reply:
I do not agree.

Reply:
Okay.  So test it.

Subscribe to Foo. (call it Sub 1)
Send a Foo message with RETAIN=1 (for the new subscription) - aka. Msg 1

Create the subscription record for another Foo subscription (Sub 2) without doing the actual subscription (simulating debounce - https://github.com/home-assistant/core/pull/88862)
Send a Foo topic message with RETAIN=0 (simulating a message sent that the original subscriber would receive) - aka. Msg 2
Do the actual subscription for Foo again (simulating the debounce actually performing the subscription async)
Send a Foo message with RETAIN=1 (which would be the RETAIN=1 message sent for Sub 2 subscribing by MQTT) - aka. Msg 3

Your code would have
Sub 1 got Msg 1 and Msg 2
Sub 2 got Msg 2 only.
Nothing got Msg 3.  Sub 2 never got a RETAIN=1 message, ever.

After this, the value of self._retained_init[Sub 1] = 1, the value of self._retained_init[Sub 2] = 0

Why did this happen?  Because init_status is NOT the value of self._retained_init[X], it is the value of `self.retained_init[X] is None`.  Meaning your code doesn't say "Ignore this RETAIN=1 message if I've already received a RETAIN=1 message", it says "Ignore this RETAIN=1 message if I have received any message at all.  Regardless of it's RETAIN value."

This issue will become apparent when https://github.com/home-assistant/core/pull/88862 lands (ie. debouncing subscriptions).  And it's the wrong behavior.
Sub 2 should have gotten Msg 2 and Msg 3.  Sub 1 should get Msg 1 and Msg 2.  ie. both should receive/process at least ONE (and only one) RETAIN=1 message.

If the RETAIN=1 message says 'This is the complete status of my object' and the RETAIN=0 messages just say 'Update field X to Y', this means Sub 2 will _never_ get a complete status of the object, only field updates (which if they are relative, say 'Set field X = X + 5' it can't do anything with). 

Due to multiple subscriptions on the same topic, especially with debouncing, you can't guarantee that the first message received will be RETAIN=1.  And if the topic sends out RETAIN=1 messages, every subscriber for that topic should process ONE (and only one) of them.  The current code is a pure race condition that can/will cause some entities to not get full values, only partial updates if the publishing entity sends full values as RETAIN=1 and deltas as RETAIN=0.

Of course, if you still don't agree, once debouncing goes in, I'll just file a bug when this race condition happens.  Because with debouncing, it's pretty much guaranteed with the current implementation.

Reply:
Also - as previously mentioned - this code will completely break with wildcard entries.  Because the wildcard sub will record the first entry (say a RETAIN=1) for the wildcard topic, so, let's say Sub 3 is for 'foo/#/config', when it receives the first message, say, 'foo/A/config', it will set:
self._retained_init[Sub 3] = 1
Now when it gets 'foo/B/config', which is ALSO a RETAIN=1 message (because wildcard subscriptions can result in one RETAIN=1 message PER matching topic), it will drop it on the floor.

Reply:
These tests prove you are correct for non-wildcards, but wildcards are broken.  Do with it what you will.

```
diff --git i/tests/components/mqtt/test_init.py w/tests/components/mqtt/test_init.py
index 5c9e3449ce..60295829ec 100644
--- i/tests/components/mqtt/test_init.py
+++ w/tests/components/mqtt/test_init.py
@@ -881,6 +881,147 @@ async def test_subscribe_topic(
         unsub()
 
 
+async def test_delayed_subscribe_topic_with_retain(
+    hass: HomeAssistant,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
+    calls: list[ReceiveMessage],
+) -> None:
+    """Test the subscription of a topic."""
+
+    calls1 = []
+    def record_calls1(*args):
+        """Record Calls."""
+        calls1.append(args[0])
+
+    calls2 = []
+    def record_calls2(*args):
+        """Record Calls."""
+        calls2.append(args[0])
+
+    await mqtt_mock_entry_no_yaml_config()
+    unsub1 = await mqtt.async_subscribe(hass, "test-topic", record_calls1)
+
+    # Initial RETAIN=1 message (for sub1)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 1", 0, True)
+    await hass.async_block_till_done()
+
+    # Since we don't care if the mqtt sub actually happened, we can pretend it was delayed
+    # until after test-payload 2 has been received.
+    unsub2 = await mqtt.async_subscribe(hass, "test-topic", record_calls2)
+
+    # A RETAIN=0 message (something happened on the topic)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 2", 0, False)
+    await hass.async_block_till_done()
+
+    # Initial RETAIN=1 message (for sub2)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 3", 0, True)
+    await hass.async_block_till_done()
+
+    assert len(calls1) == 2
+    assert calls1[0].topic == "test-topic"
+    assert calls1[0].payload == "test-payload 1"
+    assert calls1[0].retain == True
+    assert calls1[1].topic == "test-topic"
+    assert calls1[1].payload == "test-payload 2"
+    assert calls1[1].retain == False
+
+    assert len(calls2) == 2
+    assert calls2[0].topic == "test-topic"
+    assert calls2[0].payload == "test-payload 2"
+    assert calls2[0].retain == False
+    assert calls2[1].topic == "test-topic"
+    assert calls2[1].payload == "test-payload 3"
+    assert calls2[1].retain == True
+
+    unsub1()
+    unsub2()
+
+
+async def test_delayed_subscribe_topic_without_retain(
+    hass: HomeAssistant,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
+    calls: list[ReceiveMessage],
+) -> None:
+    """Test the subscription of a topic."""
+
+    calls1 = []
+    def record_calls1(*args):
+        """Record Calls."""
+        calls1.append(args[0])
+
+    calls2 = []
+    def record_calls2(*args):
+        """Record Calls."""
+        calls2.append(args[0])
+
+    await mqtt_mock_entry_no_yaml_config()
+    unsub1 = await mqtt.async_subscribe(hass, "test-topic", record_calls1)
+
+    # Initial RETAIN=1 message (for sub1)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 1", 0, False)
+
+    await hass.async_block_till_done()
+
+    # Since we don't care if the mqtt sub actually happened, we can pretend it was delayed
+    # until after test-payload 2 has been received.
+    unsub2 = await mqtt.async_subscribe(hass, "test-topic", record_calls2)
+
+    # A RETAIN=0 message (something happened on the topic)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 2", 0, False)
+    await hass.async_block_till_done()
+
+    # Initial RETAIN=1 message (for sub2)
+    async_fire_mqtt_message(hass, "test-topic", "test-payload 3", 0, False)
+    await hass.async_block_till_done()
+
+    assert len(calls1) == 3
+    assert calls1[0].topic == "test-topic"
+    assert calls1[0].payload == "test-payload 1"
+    assert calls1[0].retain == False
+    assert calls1[1].topic == "test-topic"
+    assert calls1[1].payload == "test-payload 2"
+    assert calls1[1].retain == False
+    assert calls1[2].topic == "test-topic"
+    assert calls1[2].payload == "test-payload 3"
+    assert calls1[2].retain == False
+
+    assert len(calls2) == 2
+    assert calls2[0].topic == "test-topic"
+    assert calls2[0].payload == "test-payload 2"
+    assert calls2[0].retain == False
+    assert calls2[1].topic == "test-topic"
+    assert calls2[1].payload == "test-payload 3"
+    assert calls2[1].retain == False
+
+    unsub1()
+    unsub2()
+
+
+async def test_subscribe_wildcard_topic(
+    hass: HomeAssistant,
+    mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
+    calls: list[ReceiveMessage],
+    record_calls: MessageCallbackType,
+) -> None:
+    """Test the subscription of a topic."""
+    await mqtt_mock_entry_no_yaml_config()
+    unsub = await mqtt.async_subscribe(hass, "foo/#", record_calls)
+
+    async_fire_mqtt_message(hass, "foo/A", "test-payload", 0, True)
+    async_fire_mqtt_message(hass, "foo/B", "test-payload", 0, True)
+
+    await hass.async_block_till_done()
+    assert len(calls) == 2
+    assert calls[0].topic == "foo/A"
+    assert calls[0].payload == "test-payload"
+    assert calls[0].retain == True
+    assert calls[1].topic == "foo/A"
+    assert calls[1].payload == "test-payload"
+    assert calls[1].retain == True
+
+    unsub()
+
+
 async def test_subscribe_topic_non_async(
     hass: HomeAssistant,
     mqtt_mock_entry_no_yaml_config: MqttMockHAClientGenerator,
```

Reply:
Please stop there's text walls, Again, I do not agree.

Reply:
You were right about the wild card handling.

Reply:
> Please stop these text walls, Again, That does not help. I'll have a look at the wild cards.

Oh, sorry for making you read a full explanation.  I don't work in 'sound bytes', I try and explain myself.  So that when I say something, nobody can claim I made a snap decision or didn't think about the problem (like, say, closing things as "it's zigbee2mqtt's problem" without looking at the evidence to the contrary provided ... or just saying "I don't agree" until I actually prove it with unit tests you couldn't be bothered writing).  They can see my reasoning/work - which is why I pasted the proof (the 'wall of text' is a diff with unit tests).  I've found that's something most good engineers do.

But you're right.   Reading hard.  Proper evidence or explanations are fake news.  No instant gratification.  I'll go back to watching Fox News now.

Man I'll be glad when these bugs are fixed and I don't have to deal with this component again.